Integrate splain – implicit resolution chains and type diffs in error messages

tek opened the pull request:
This integrates https://github.com/tek/splain into the compiler

* `-Vimplicits` makes the compiler print implicit resolution chains when no implicit value can be found
* `-Vtype-diffs` turns type error messages (`found: X, required: Y`) into colored diffs between the two types

I copied most of splain's code into the directory `src/compiler/scala/tools/nsc/typechecker/splain` and changed its integration to be less intrusive. There are now compiler options like `-Ysplain` to control its behaviour. I kept the analyzer plugin method that allows consumers to override or amend output formatting, e.g. for providing more detailed messages about specific typeclasses with access to the typechecker and involved types.

 (previous PR at https://github.com/scala/scala/pull/5958).

tek committed:
Integrate splain – implicit resolution chains and type diffs in error messages

This error formatting framework displays a tree of implicit parameters
that correspond to the resolution chain between an error's call site and
the offending implicit.

Several additional improvements for error formatting are provided as
well, like colored diffs of found/required error types, which are based
on a set of pure data types extracted from the compiler internals,
available to plugin developers through the AnalyzerPlugin API.

som-snytt commented:
You might consider grouping the compiler flags, `-Ysplain:no-color, trunc-refined` etc.

tek commented:
@som-snytt could you elaborate how that would work? The only tool I'm seeing that allows grouping like this is `MultiChoiceSetting`, but it doesn't seem very suitable for this specific case

som-snytt commented:
It seems obvious to me; only the int-valued settings can't be grouped this way (although one could imagine supporting `-Ysplain:myint=3` syntax).

The other possibility for [avoiding a huge `-Y` help page](https://github.com/scala/scala/pull/7686) would be to only show prefixes under `-Y` and then `-Ysplain:help` would show options with that prefix.

som-snytt commented:
Note that you could also provide a man page for: `man splain`.

tek commented:
ok, got it!
I hoped that specifying only `-Ysplain` would make `setByUser` true, but it's gonna have to be `-Ysplain:enable`.

julienrf commented on test/files/run/splain.check:
```diff
@@ -0,0 +1,22 @@
+newSource1.scala:13: error: implicit error;
+!I e: II
+ImplicitChain.g invalid because
+!I impPar3: I1
```
I find this message a bit hard to interpret. I have the impression that Dotty provides a more helpful error message ([see an example here](https://scastie.scala-lang.org/c2BsN4fuT6ebCImGdSQBQA)). Am I right, or are there limitations in what Dotty does that are addressed by splain?

tek replied:
the motivation for this notation is to be terse, concise and parsable in the face of implicit chains that are tens of levels deep. I haven't gotten any feedback on this by splain users so far, so I can't tell whether this is as useful to others as it is to me.
The test output with dummy names is certainly less readable than one where the involved names and types are recognizable, like `circe.Decoder` and `shapeless.Generic`, which is the intended use case.

One solution would certainly be to offer both verbose and terse messages, to be configured by the user.

Are you suggesting that Dotty's messages would make splain obsolete? Would that imply that it wouldn't be needed in current Scala?
I haven't looked at Dotty's implicit messages yet, does it display whole chains?

julienrf replied:
> Are you suggesting that Dotty's messages would make splain obsolete?
> Would that imply that it wouldn't be needed in current Scala?

This is not clear yet. In any case, it’s definitely useful to work on improving error messages!

> I haven't looked at Dotty's implicit messages yet, does it display whole chains?

I have posted a comparison of messages produced by Scala and Dotty [here](https://contributors.scala-lang.org/t/better-implicit-search-errors-problematic-cases-wanted/3587/2)

tek replied:
I will have to investigate some test cases myself.

In any case, implicit conversions are not a feature that is handled by splain explicitly, just FYI.

My question now is: do you want to make this PR about how error messages should look? So far the mission was to make splain available as-is – if this should be different, we should make this an explicit statement.

julienrf replied:
> My question now is: do you want to make this PR about how error messages should look? So far the mission was to make splain available as-is – if this should be different, we should make this an explicit statement.

That’s a good point. I agree that working on how error messages should look could be done in another PR.

SethTisue replied:
marking as "resolved" because although further improvement would be welcome, it's not a blocker

kubukoz commented:
Hi, what's the state of this PR? It's been some time since the last update :)

tek commented:
I am waiting for activity!

som-snytt commented:
With the new github inbox for notifications, I'll never miss another review request.

I'll get up to speed. First, I'll get up.

som-snytt commented:
With the new github inbox for notifications, I'll never miss another review request.

I'll get up to speed. First, I'll get up.

SethTisue commented:
@lrytz we're open to eventually merging this, right? I think you and I discussed it a few weeks ago but I admit that I forgot what we said. did we have any medium-sized-or-larger reservations about it?

julienrf commented:
FWIW, I haven’t had a look at the PR implementation but I think the feature itself is extremely valuable.

julienrf commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -524,4 +524,49 @@ trait ScalaSettings extends AbsScalaSettings
+
+  val Ysplain: MultiChoiceSetting[YsplainChoices.type] =
+    MultiChoiceSetting(
+      name = "-Ysplain",
```
What do you think of a name like `-Yexplain-implicits` instead? (we already have `-Ylog-implicits`)

tek replied:
I have no reservations about getting rid of the branding. whatever sounds good to the team.

som-snytt replied:
Actually `-Vimplicits` in 2.13. Maybe two options is not necessary. But it would be cool to keep `-Ysplain` as an alias.

kubukoz replied:
Let's pick one :) I think `-Yexplain-implicits` is better because users new to splain won't know it's about implicits at all if they see the option in a build.

SethTisue replied:
Som's right that it should be a `-V` something something, though, isn't he?

kubukoz replied:
`-V` is for "verbose", right? In that case, maybe `-Vimplicit-errors`?

tek replied:
I'm confused now! Should it be `-Vexplain-implicits`?

tek replied:
oops, didn't see that last comment

kubukoz replied:
Let's just pick one ;)

martijnhoekstra replied:
vexplain is my new favourite word.

som-snytt replied:
I don't mind `-Vimplicits`. The simplified options avoid compounds that are hard to remember or discover. (I could see `-Vexplain` with `-Vexplain:types,implicits` and retire `-explaintypes`. But "verbose" is redundant with "explain".)

tek replied:
So should we start with `-Vimplicits` and integrate it with `-explaintypes` later?

SethTisue replied:
marking as resolved because this discussion is now stale (there is further discussion on this in the December round of feedback) and because the issue is on our March-2021 list of blockers

lrytz commented:
> @lrytz we're open to eventually merging this, right? I think you and I discussed it a few weeks ago but I admit that I forgot what we said. did we have any medium-sized-or-larger reservations about it?

I also don't remember :-)

My only reservation here is that the improvements live behind new compiler flags, and that the implementation is separate from the ordinary error messages and type printing. Can we integrate this deeper and make more people benefit from the helpful messages?
  - for infix type printing, we have `@showAsInfix` - how does it relate?
  - type diffs would be really good to have by default, is there a reason against it? maybe not to break IDEs?
  - can the implementation of type printing be combined with what already exists in the compiler?
  - is there a drawback in adding compact implicit chains by default? compiler perfromance?

SethTisue commented:
> My only reservation here is that the improvements live behind new compiler flags

Even if we decide that this style of output should remain under a flag, perhaps one or more normal error messages could mention the flag, to improve discoverability.

Regardless, I agree that deeper integration is better. Because passing flags is cumbersome; people tend to either turn a flag on all the time, or never.

kubukoz commented:
Frankly I'd rather have it soon as a flag than wait months and get new behavior as a default in a 2.13.x version :)

lrytz commented:
@kubukoz that's understandable of course :-)

However, features we add to the compiler are shipped to the entire Scala community, so changing things (even compiler flags or the style of error messages) comes with a cost, that's why things often move a bit slower.

Also, from our perspective, it's a piece of code that we'll have to maintain for a long time, so it makes sense to integrate well into our codebase. For example, `showType` is re-implementing functionality that exists (to some degree) in the compiler.

tek commented:
I would suggest starting out with the flag, since the plugin has so far only been used by people intentionally looking for the provided functionality, and some of the edge cases in, e.g. the found/req type diffs can be slightly buggy – not to the point of significant impediment, but it could probably need a bit of bug reporting.

There's nothing keeping us from making it the default in the future, and it does feel like a lot of commitment to do it out of the gate.

lrytz commented:
35% of the respondents in https://scalacenter.github.io/scala-developer-survey-2019/ say that "handling missing implicit errors" is a "main pain point in daily workflow".

tek commented:
fair point :smile_cat: 

regarding your concern about integrating it deeper: One benefit of having it separated, especially with Analyzer Plugin machinery involved, is that it would make it very simple to provide *highly* customized error message plugins for specific libraries *with access to the typer*. One example of this is the builtin shapeless Record and Lazy support.

Of course there is some compromise in between. My only concern is that this PR and its predecessor has been going on for years and it would be nice to close the plugin repo and develop this in the compiler with some version already released, because it would make development significantly simpler. A flag seems a good approach, treating it as an experimental feature.

tek commented:
So my proposal would be 2.13.4 with `-Vimplicits`, 2.13.5 as default

lrytz commented:
> fair point 😸

Not sure what point I made 😀 I definitely support better diagnostics, and the survey shows that there's a real need.

I'm a bit torn about integrating it in the current form, as I expressed before. @SethTisue promised to also share his thoughts.

tek commented:
so what exactly would deeper integration entail in order to make this ready for merge?

SethTisue commented:
I'll say more later but I'll briefly say just now that the separate implementation of type printing bugs me, it seems like tech debt we shouldn't take on. It's all to easy to say "oh just throw it in, it's experimental" but you can't just wave an "experimental" wand at the community, as soon we ship something than people start using and it and expecting it regardless of what we said about its status, and any tests that regress will still fail PR validation and block releases, etc etc.

I can say more but that's the 30 second version.

tek commented:
by "separate implementation of type printing", you mean that in messages that are not handled by splain there is a separate output formatting routine in use?

SethTisue commented:
> by "separate implementation of type printing", you mean that in messages that are not handled by splain there is a separate output formatting routine in use?

Exactly. so e.g. `formatNormalSimple` seems to me like code we shouldn't be adding. Perhaps I simply don't appreciate the motivation for why this reimplementation was done — in that case, comments should explain the motivation.

tek commented:
There are truncation rules applied in `formatNormalSimple`, the motivation to have this separate is simply that it was intended as an optional simplification. What would you suggest to be done with this? Should it replace the regular printer?
I will take another look but I think this is implemented object-oriented, and I doubt it would be possible to combine the two easily.

SethTisue commented:
> I can say more but that's the 30 second version.

I have now taken the time to actually read the diffs, which I hadn't done before.

Some additional remarks:

* There is some nice documentation in the https://github.com/tek/splain repo that I think could be made part of this PR, so that it then shows up in our autogenerated compiler options doc at https://docs.scala-lang.org/overviews/compiler-options/index.html . (But see also my remarks below about whether all this configurability is actually good to provide or not.) If it's too long/unwieldy for that page, there could be a separate page on https://docs.scala-lang.org/overviews/index.html.
* The amount of testing seems rather light. I mean both that the core functionality gets only a handful of testing, and that there are a number of configuration options that don't seem to have any test coverage at all.
* How many of the configuration options really needed/wanted? Might some of them be experiments you did that we should decide either panned out, or didn't? The more options, the more it costs users to study and choose between them, and the more it costs us as maintainers to document, provide adequate test coverage for, and maintain indefinitely.
* I have doubts about including the hardcoded support for `shapeless.labelled.KeyTag` and `shapeless.tag.Tagged`. It just seems strange for the compiler to have that kind of dependency on third-party code. (And of course "strange" isn't only just aesthetics, but a matter of maintenance/testing/docs.)

tek commented:
`scala.reflect.internal.Type` has `def safeToString` which is invoked for error messages via `toString` at the site of the error. I would assume that this would have to be changed in hundreds of locations to make splain the default.

Remarks:
1. sounds good!
2. no doubt. I will add those once we have reached some more consensus.
3. I think all of them are helpful, but I have no idea how popular they are. Some were implemented in response to a feature request. I'd be happy to go over them to decide which ones should be removed, hardcoded or kept. Some of them are just for improving readability of long types. I'm not sure how to apply Occam's Razor here, it's my first work on a compiler :smile: 
4. Absolutely, that's where the plugin concept would come in that I mentioned previously. I would remove those before merging.

tek commented:
@SethTisue update:

1) I threw together a preliminary documentation page at https://github.com/tek/docs.scala-lang/commit/a31fd77b3d52fa6b48a22b8d0b5efea70841cd2d , please have a look!
2) Pending.
I'm going to copy the rest of the tests from the plugin repo and come up with some additional ones.
3) I removed most of the extra options, now there are only three left.
I think `no-color` and `verbose-tree` is pretty essential. About `trunc-refined` I'm inconclusive, not sure this should be done by default but also unsure if this passes the significance test. Long refined types in implicits are often quite noisy, but I don't know how frequently that is a problem.
I hardcoded `break-infix` to use 70 characters.
4) Removed the shapeless stuff.

tek commented:
ok, I changed the option to `-Vimplicts` and just now realized that that is the new name of `-Xlog-implicts`.

tek commented:
it's now `-Vexplain-implicits`.

I would however prefer something more general, since not all features pertain to implicits (like foundreq diff).
Like `-Verrors`.

tek commented:
@SethTisue @kubukoz @lrytz RFC!

kubukoz commented:
I'm good with the current name :)

tek commented:
@kubukoz how do you like the documentation page?

kubukoz commented:
> has been introduced in Scala 2.14

This doesn't look right, unless I'm behind on the version schedule :)

tek commented:
if we manage to get this done before 2.14, we can still change it :smile:

kubukoz commented:
I was under the impression 2.14 isn't happening (before dotty) 😅

tek commented:
really? in any case, I will substitute the appropriate version before go live.

SethTisue commented:
re: 2.14, https://www.scala-lang.org/2019/12/18/road-to-scala-3.html

(sorry but I won't be able to look at this PR again until after 2.13.2 is out)

som-snytt commented:
What month is it? Today feels like Wednesday. I'll take a look and follow up with status. I remember when we used to call vanilla flavor "plain" so this feature could be called "svanilla".

som-snytt commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -551,4 +551,32 @@ trait ScalaSettings extends StandardScalaSettings with Warnings {
+
+  val VimplicitsTruncRefined: IntSetting =
+    IntSetting(
+      "-explain-VexplainImplicits-trunc-refined",
```
This option name looks like a typo or late-night edit. Consider `-Y` to modify behavior of `-Vexplain-implicits`. `-Yexplain-implicits-trunc-refined` or something. Or just `-V`, not a big deal. I was thinking: easier autocomplete perhaps, and `-Y` is less visible or stable. There are a couple of "show trees" modifier options that are `-Y` on this basis.

som-snytt commented:
Not a deal breaker, just to document that the long text is in parens and the very end looks like a method signature.

```
scala> Future(42)
             ^
       error: implicit error;
       !I executor (Cannot find an implicit ExecutionContext. You might pass
       an (implicit ec: ExecutionContext) parameter to your method.

       The ExecutionContext is used to configure how and on which
       thread pools Futures will run, so the specific ExecutionContext
       that is selected is important.

       If your application does not define an ExecutionContext elsewhere,
       consider using Scala's global ExecutionContext by defining
       the following:

       implicit val ec: scala.concurrent.ExecutionContext = scala.concurrent.ExecutionContext.global): ExecutionContext
```

som-snytt commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -551,4 +551,32 @@ trait ScalaSettings extends StandardScalaSettings with Warnings {
+      helpArg = "feature",
+      descr = "print dependent missing implicits",
+      domain = VimplicitsChoices,
+      default = None,
```
`default = Some("enable" :: Nil),` to allow `-Vexplain-implicits`.

TIL you can `-Vexplain-implicits enable file.scala`, so I guess it only consumes valid option args! Clever!

tek replied:
nice!!!

som-snytt commented on src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:
```diff
@@ -179,6 +179,15 @@ trait AnalyzerPlugins { self: Analyzer =>
+    /**
+     * Construct a custom error message for implicit parameters that could not be resolved.
+     *
+     * @param tree The tree that requested the implicit
```
Stale doc.

som-snytt replied:
Luckily no one reads docs.

tek replied:
fixed

som-snytt commented on test/files/run/splain.scala:
```diff
@@ -0,0 +1,53 @@
+object Test
+extends DirectTest
+{
+  override def extraSettings: String = "-usejavacp -Vexplain-implicits:enable -Vexplain-implicits:no-color"
```
or `-Vexplain-implicits enable no-color` etc. That makes me think `enable` should be set if the other options are set, so `-Vexplain-implicits:no-color` should suffice.

som-snytt replied:
That's possible with `.withPostSetHook(s => )` but I'm too lazy to work out the semantics. But could be added later if worth it.

tek replied:
I gave this a try, please tell me what you think @som-snytt

som-snytt reviewed (changes requested):
Just the weird `-option-option` name to fix. And default to `enable`.

Maybe we can just push a fix. I'll experiment a bit more, but it should get a nightly workout.

som-snytt commented:
I'll push a fix to my comment. I just re-read the pre-covid comments here. Questions about integration and DRY and test coverage are still germane but maybe not blockers for moving forward. I don't have time to donate to high gloss. I did switch the "max refined" setting to `-Y`, since who knows if it even pulls its weight (as commented above).

som-snytt reviewed (approved):
Very nice follow-up! :shipit: or at least release to intrepid adopters of nightlies.

tek commented:
> Not a deal breaker, just to document that the long text is in parens and the very end looks like a method signature because of the rparen. (Due to crosstalk with the example code.)

I modified the output to print the annotation message starting on the next line. please take a look!

![image](https://user-images.githubusercontent.com/2632/89546674-c11f3600-d7fc-11ea-8c48-1c1c75b9b969.png)

tek commented:
there are a few improvements in the compiler plugin project that need to be backported, should we do that after merging?

SethTisue commented:
I'm sorry I've let this sit for so long. It won't make 2.13.4, but I will return my attention to it for 2.13.5.

dwijnand commented:
Last two commits need squashing to get all the commits passing.

som-snytt commented:
Please update the PR title to say, "splain, take 3"...  Oh, that reminds me, maybe we'll see it in Scala 3 first?

tek commented:
:smile:

SethTisue commented:
@tek if you rebase this one more time, it will be ready for merge for 2.13.5, afaics. (yay!)

before 2.13.5 is released, we'll need the PR to have a good description of this for users, since it will be linked from the release notes.

> there are a few improvements in the compiler plugin project that need to be backported, should we do that after merging?

hmm, well, you tell us, how important are they?

tek commented:
@SethTisue 

> if you rebase this one more time

there seems to be no 2.13.4 branch, is 2.13.x correct for this?
 
> before 2.13.5 is released, we'll need the PR to have an informative title and a good description of this, for users, since it will be linked from the release notes.

Which description do you mean, the first comment of the PR?
There's a PR in the doc repo as well.
 
> hmm, well, you tell us, how important are they?

I wouldn't call them important, just treatment of a few special cases, like refined type. But could be a bit of a hassle to merge this now.

SethTisue commented:
> there seems to be no 2.13.4 branch, is 2.13.x correct for this?

yes

> Which description do you mean, the first comment of the PR?

yes

> There's a PR in the doc repo as well

if there is good doc there, then the PR description could just link to it

SethTisue commented:
> I wouldn't call them important, just treatment of a few special cases, like refined type. But could be a bit of a hassle to merge this now.

okay, sounds like it could be one or more later, separate PRs

> Should I squash all commits now? I squashed the last two to address @dwijnand 's concern, but it doesn't seem to have had the desired effect.

Yes please. Every commit must pass CI. More commits may facilitate understanding the history; squashing into fewer commits means less work for you.

tek commented:
Should I squash all commits now? I squashed the last two to address @dwijnand 's concern, but it doesn't seem to have had the desired effect.

SethTisue commented:
> Should I squash all commits now? I squashed the last two to address @dwijnand 's concern, but it doesn't seem to have had the desired effect.

Every commit must pass CI. More commits may facilitate understanding the history; squashing into fewer commits means less work for you. Squashing the entire thing to a single commit can be fine, depending on the nature of the changes.

tek commented:
> if there is good doc there (or wherever), then the PR description could just link to it

looks like I haven't created a PR, just a commit. will do that now.

SethTisue commented:
I just reviewed the diffs and a single commit seems appropriate, actually.

Keeping it as separate commits is helpful during the review process, but the review process is pretty much over now, so (for this PR in particular, anyway) I think it's not necessary to preserve the history.

tek commented:
I agree!

SethTisue commented:
@lrytz I think your concerns were addressed, but you might want to take one last look?

lrytz commented on src/compiler/scala/tools/nsc/typechecker/splain/SplainErrors.scala:
```diff
@@ -0,0 +1,60 @@
+    targs: List[Type],
+    tparams: List[Symbol],
+    originalError: Option[AbsTypeError],
+  ): Unit = {
```
`if (settings.implicitsSettingEnable)`?

lrytz commented on src/compiler/scala/tools/nsc/typechecker/splain/SplainErrors.scala:
```diff
@@ -0,0 +1,60 @@
+        case _ => ()
+      }
+    }
+    (implicitTree: @unchecked) match {
```
`if (settings.implicitsSettingEnable)`?

lrytz commented on src/compiler/scala/tools/nsc/typechecker/splain/SplainData.scala:
```diff
@@ -0,0 +1,107 @@
+
+    def removeErrorsFor(tpe: Type): Unit = errors = errors.dropWhile(_.tpe == tpe)
+
+    def startSearch(expectedType: Type): Unit = {
```
`if (settings.implicitsSettingEnable)`?

tek replied:
are you concerned about performance impact?

lrytz replied:
Yeah, cc @retronym

lrytz commented on src/compiler/scala/tools/nsc/typechecker/splain/SplainData.scala:
```diff
@@ -0,0 +1,107 @@
+      stack = expectedType :: stack
+    }
+
+    def finishSearch(success: Boolean, expectedType: Type): Unit = {
```
`if (settings.implicitsSettingEnable)`?

lrytz commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -575,4 +575,39 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
+    MultiChoiceSetting(
+      name = "-Vexplain-implicits",
+      helpArg = "feature",
+      descr = "print dependent missing implicits",
```
Can we make this a little more explanatory? It has to stay a one-liner though, as this is what shows up in `scalac -V`.

Sadly, `MultiChoiceSetting` doesn't have a `helpText` parameter. Maybe change it to `MultiStringSetting`? See [`Wconf`](https://github.com/scala/scala/blob/95cb2592aac26375e67e772fc438013e95af8e60/src/compiler/scala/tools/nsc/settings/Warnings.scala#L33) for example. I think the implicit trace syntax really should be explained, or there should be a link to an explanation, I personally cannot "just" make sense of it when I see it.

tek replied:
would be a link to the page on docs.scala.org be satisfactory?

lrytz replied:
IMO that would work, @SethTisue what do you think?

SethTisue replied:
yes, I think that would be fine

SethTisue replied:
hitting "resolve" since it's now on our list of checkboxes

lrytz commented on test/files/run/splain.check:
```diff
@@ -0,0 +1,62 @@
+newSource1.scala:13: error: implicit error;
```
I find `implicit error` very terse. Also the default message says upfront what the target parameter is. Could we keep `could not find implicit value for parameter X: T`?

SethTisue replied:
marking as "resolved" only because it's not a blocker; would still be good to see this addressed in a followup

lrytz commented on test/files/run/splain.check:
```diff
@@ -0,0 +1,62 @@
+  implicitly[II]
+            ^
+newSource1.scala:6: error: type mismatch;
+  L|R
```
Compared to the default message
```
 found   : FoundReq.L
 required: FoundReq.R
```
this doesn't contain the prefix. I'm not sure exactly what the logic is in the default messages, what prefixes are kept. Maybe worth taking a look?

tek replied:
there is a lot of disambiguation logic and configuration options for prefixes in the plugin, though some of that is still outstanding for backport (see conversation with @SethTisue above) or has been removed (it was requested to cut down the number of options somewhere).

So this is already on the agenda, we can add options back later if we deem them essential!

lrytz replied:
👍

lrytz commented on test/files/run/splain.check:
```diff
@@ -0,0 +1,62 @@
+
+  implicitly[F[Arg]]
+            ^
+newSource1.scala:4: error: implicit error;
+!I ec: ExecutionContext
+  Cannot find an implicit ExecutionContext. You might add
```
In the base case when there were no discarded alternatives, the `!I ...` line doesn't seem to add any value?

Also, the default message skips the `could not find implicit value` part (`implicit error` here) when the target has an `@implicitNotFound`.

SethTisue replied:
marking as resolved since improvement here would be welcome but it's not a blocker

lrytz reviewed (commented):
`-Vexplain-implicits` also enables type diff printing in found/req, which is not about implicits. Should we rename the option, or introduce a separate option for type diff printing?

How does the new option compare to the existing `-Vimplicits`? Does it make sense to keep it?

Suggestion: it would be valueable to run `partest --neg --update-check` with the feature enabled and look at the diff.

I also added a few comments about message formatting, but I agree with previous discssions that these changes can come as follow-ups.

tek commented:
> `-Vexplain-implicits` also enables type diff printing in found/req, which is not about implicits. Should we rename the option, or introduce a separate option for type diff printing?

There used to be an option, but I removed it at your request! (You said "activate by default", which suggests it should not be handled by `-Vexplain-implicits`, so the flag could be reintroduced and default to enabled)

> How does the new option compare to the existing `-Vimplicits`? Does it make sense to keep it?

I would say no, since I've basically started this project because `-Vimplicits` was inadequate, but I'm not sure whether splain is a subset of what `Vimplicits` shows.

> I also added a few comments about message formatting, but I agree with previous discssions that these changes can come as follow-ups.

just saw this after responding to the comments with basically the second part of the sentence, so I deleted the response! :sweat_smile:

lrytz commented:
> There used to be an option, but I removed it at your request! (You said "activate by default", which suggests it should not be handled by `-Vexplain-implicits`, so the flag could be reintroduced and default to enabled)

Looking at the code, I'm now a little worried about enabling type diffs by default, because that replaces the current type-to-string implementation with a different one, and we get changes like prefixes not being shown. You wrote yourself "the found/req type diffs can be slightly buggy".



> I would say no, since I've basically started this project because `-Vimplicits` was inadequate, but I'm not sure whether splain is a subset of what `Vimplicits` shows.

If we can agree on that I think it's best to just reuse `-Vimplicits`. I wouldn't worry too much about it, before 2.13 the option was in the `-X` category.

tek commented:
> Looking at the code, I'm now a little worried about enabling type diffs by default, because that replaces the current type-to-string implementation with a different one, and we get changes like prefixes not being shown. You wrote yourself "the found/req type diffs can be slightly buggy".

So is "active by default, but only with `-Vimplicts`" not a sufficient compromise?
And if not, should the new option be orthogonal to `Vimplicits`?
Any suggestions for a category?

> If we can agree on that I think it's best to just reuse `-Vimplicits`. I wouldn't worry too much about it, before 2.13 the option was in the `-X` category.

would it be fine if the people in this PR made the decision to remove that option? I couldn't find who made that change in the commit history

SethTisue commented:
Lukas has pointed out that we never fully resolved the question of type printing. 

@tek you wrote above, in defense of keeping the custom type printing code:

> There are truncation rules applied in formatNormalSimple, the motivation to have this separate is simply that it was intended as an optional simplification. What would you suggest to be done with this? Should it replace the regular printer? I will take another look but I think this is implemented object-oriented, and I doubt it would be possible to combine the two easily.

I'm sorry but I don't understand what "the motivation to have this separate is simply that it was intended as an optional simplification" means. Can you spell out what you mean by that?

> I would assume that this would have to be changed in hundreds of locations to make splain the default.

I believe we are suggesting that you use the existing type printing in the compiler, rather than replacing the existing code with yours.

If that doesn't make sense, can you help me understand why?

tek commented:
@SethTisue I don't have the entire thing in my head at the moment, but to start from the top: The primary reason for the entire printing machinery is to get the type diffs. As far as I can tell, it is consensus that this should be part of the PR.

For this purpose, the types are converted to a set of pure data types, which are then processed by the diff printer. Doing this with the native types would result in much less maintainable code, I would imagine.

Printing the entire type with the builtin printer would circumvent the diffs, so this would be incompatible. Are we on the same page here?

If you're only talking about the string-returning "leaf" functions like `formatNormalSimple`, I'd need more information about which parts exactly should use the type printer (and optimally, how :sweat_smile: ). In this function, for example, the refinement type truncation feature is applied, which could certainly be removed, but is that desired? So I speculate that there is very little that is reimplemented without necessity, but I'll gladly be convinced otherwise!

PS: I don't remember what I meant with "optional simplification".

retronym commented:
> For this purpose, the types are converted to a set of pure data types, which are then processed by the diff printer. Doing this with the native types would result in much less maintainable code, I would imagine.

Could those pure data types be the AST nodes that represent syntactic types? That would be a useful utility to have in the compiler codebase. I have some unmerged code lying around that does this in https://github.com/scala/scala/compare/2.13.x...retronym:topic/scalap-lite-2

tek commented:
@retronym those types encode mainly formatting aspects, so I'd assume it wouldn't be trivial.

kubukoz commented:
Can we pick one way to print for now, and change it later? If it's just for diagnostics, users shouldn't rely on the behavior so it should be easy to change afterwards... unless it's not 😅

tek commented:
> this didn't make 2.13.5 because we ended up rushing 2.13.5 to get a new TASTy reader out, but let's try to get this in 2.13.6.

:+1: 
 
>     * [ ]  did we settle on the name `-Vimplicits`, replacing the old functionality that flag had; does the code now reflect that decision?
>       
>       * the exchange about this ended with @tek's [#7785 (comment)](https://github.com/scala/scala/pull/7785#issuecomment-741848661) which we (Lukas and the rest of us) never responded to

It's still `-Vexplain-implicits`. Waiting for that decision!

>     * [ ]  Lukas asked about adding `if (settings.implicitsSettingEnable)` checks in a number of places, out of concern for performance; was that handled?

just pushed those changes

>     * in `ScalaSettings`, we need:
>       
>       * [ ]  the best possible one-line description
>       * [ ]  longer doc that links to the page on docs.scala.org

related: https://github.com/scala/docs.scala-lang/pull/1840

> 
>     * Lukas's comments about the wording of error messages:
>       
>       * [ ]  "I find 'implicit error' very terse", can this be easily improved, is it a blocker?
>       * [ ]  ditto on the question at [#7785 (comment)](https://github.com/scala/scala/pull/7785#discussion_r539261029)

Lukas wrote:

> I also added a few comments about message formatting, but I agree with previous discssions that these changes can come as follow-ups.

Agreed that this can be done in a follow-up?

> 
>     * [ ]  printing of type diffs: "the found/req type diffs can be slightly buggy", should we go ahead and merge anyway, or is it a blocker, or does it need assessing?

Since this feature is highly heuristical, I'd assume that there are quite a few edge cases that could cause some types to be hidden. In any case, there are some patches upstream that we can backport (later?)

tek commented:
added a link to the future documentation page in the option description

SethTisue commented:
(Dale and I will do another round tomorrow — we're going to try to keep looking at this together every day until it's done.)

tek commented:
@lrytz 

> @tek thank you. Your patience, responsiveness, rationality is really impressive!
thanks :sweat_smile: 

> > I'd assume that there are quite a few edge cases that could cause some types to be hidden
> 
> "hidden" how?

If splain decides that two types are equal, but they aren't in reality, the diff will only display one type, potentially hiding the one with the additional information (refinement etc). There is a fix upstream, though I can't say whether this handles _all_ cases, since types aren't variants of a simple ADT. I would claim though that this should happen only _very_ rarely.

> It looks like there's no way type diffs will happen otherwise. Can the PR be split up between implicit chains and type diffs? Then we could discuss that tradeoff in isolation.
Would that be much better than reintroducing the flag as opt-in?

tek commented:
added the flag `-Vtype-diffs` in case it's decided to be desired

lrytz commented:
> `-Vtype-diffs`

👍

It's a nice feature, I guess everyone agrees. I'm fine taking in the second type-to-string implementation given that it's used only for that feature. It's also safe: the existing printer is untouched, and we can take it out if we ever need to.

tek commented:
>     * [ ]  does the splain repo have test coverage that was not brought over in this PR?

added all the tests that don't depend on a removed feature (like shapeless) or were added for new fixes/features

> 
>     * [ ]  below @tek said there are recent bugfixes in the splain repo that haven't been ported here yet; decide whether to require those in this PR or ask for them in a followup PR

I'd get right on that, but I would find it sensible to get this PR merged sooner and create a followup.

som-snytt commented:
Sorry to spam with a "me too" comment, but wanted to echo Lukas that everyone has demonstrated amazing fortitude in this epic effort.

The reason everyone is going to cross-compile scala 2/3 is so that when something doesn't work in scala 3, they can enable this feature in scala 2.

dwijnand commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -507,8 +507,7 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
   val Ytyperdebug        = BooleanSetting("-Vtyper", "Trace type assignments.") withAbbreviation "-Ytyper-debug"
-  val XlogImplicits      = BooleanSetting("-Vimplicits", "Show more detail on why some implicits are not applicable.")
-    .withAbbreviation("-Xlog-implicits")
+  val XlogImplicits      = BooleanSetting("-Xlog-implicits", "Show more detail on why some implicits are not applicable.")
```
Should be removed.  Are there any usages left?  Or are they all replaced with `VexplainImplicits` usages?  Which, btw, should also be renamed so the val name matches the setting name.

tek replied:
are we sure it shouldn't stay in as `Xlog-implicits` only?

dwijnand replied:
Yes, it shouldn't stay, with the new implementation replacing it.

tek replied:
ok, then I'll remove the related tests!

tek replied:
looks like the tests can remain, just had to remove parts of the error messages

dwijnand replied:
I think we want to keep those error messages, just update them with the new implementation.  I.e. switch them to `-Vimplicits`.

tek replied:
only one of those tests has different output with `-Vimplicits` enabled, I changed it accordingly.

dwijnand commented:
> In any case, there are some patches upstream that we can backport (later?)

> added all the tests that don't depend on a removed feature (like shapeless) or were added for new fixes/features

I think we should bring the new fixes/features and tests in now, so that the implementation here isn't a fork of splain.

tek commented:
@dwijnand @SethTisue all upstream changes are now integrated!

dwijnand commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -575,4 +573,45 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
+    MultiChoiceSetting(
+      name = "-Vimplicits",
+      helpArg = "feature",
+      descr = "Print dependent missing implicits and colored found/required type diffs. See https://docs.scala-lang.org/overviews/compiler-options/errors.html",
```
> and colored found/required type diffs

No longer true, as this is behind `-Vtype-diffs` now.

dwijnand commented on src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:
```diff
@@ -107,7 +109,7 @@ trait ContextErrors {
 
     def typeErrorMsg(context: Context, found: Type, req: Type) =
-      if (context.openImplicits.nonEmpty && !settings.XlogImplicits.value && currentRun.isScala213)
+      if (context.openImplicits.nonEmpty && currentRun.isScala213)
```
Revert, except using Vimplicits now.

lrytz replied:
I agree, this looks like it should stay the same.

dwijnand commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -443,8 +445,6 @@ trait Implicits {
 
     @inline final def failure(what: Any, reason: => String, pos: Position = this.pos): SearchResult = {
-      if (settings.XlogImplicits)
-        reporter.echo(pos, s"$what is not a valid implicit value for $pt because:\n$reason")
```
Keep?

lrytz replied:
This were the old style traces which are now handled by splain, no?

dwijnand replied:
That's what I was hoping but not what the tests seem to show.  We lose these bits of information:

```
implicit-shadow.scala:6: <i2s: error> is not a valid implicit value for Int(1) => ?{def isEmpty: ?} because:
reference to i2s is ambiguous;
it is imported twice in the same scope by
import C._
and import B._
```

```
t6323a.scala:13: scala.reflect.api.`package`.materializeTypeTag[Test](scala.reflect.runtime.`package`.universe) is not a valid implicit value for reflect.runtime.universe.TypeTag[Test] because:
failed to typecheck the materialized tag:
cannot create a TypeTag referring to class Test.Test local to the reifee: use WeakTypeTag instead
      val value = u.typeOf[Test]
                          ^
```

```
implicit-log.scala:61: byVal is not a valid implicit value for Int(7) => ?{def unwrap: ?} because:
incompatible: (x: 7): 7 does not match expected type Int(7) => ?{def unwrap: ?}
    val res = 7.unwrap() // doesn't work
              ^
```

dwijnand commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -1508,17 +1510,15 @@ trait Implicits {
       )
       // todo. migrate hardcoded materialization in Implicits to corresponding implicit macros
       val materializer = atPos(pos.focus)(gen.mkMethodCall(TagMaterializers(tagClass), List(tp), if (prefix != EmptyTree) List(prefix) else List()))
-      if (settings.XlogImplicits) reporter.echo(pos, "materializing requested %s.%s[%s] using %s".format(pre, tagClass.name, tp, materializer))
```
Keep?

lrytz replied:
Use `settings.debug` here?

dwijnand commented on test/files/neg/implicit-any2stringadd.scala:
```diff
@@ -1,4 +1,4 @@
-// scalac: -Xsource:3 -Xlog-implicits
+// scalac: -Xsource:3
```
Revert, with -Vimplicits.  And the same with the rest of the tests.

SethTisue commented:
remaining loose ends:

naming, scope, and docs:

* [ ] do we like the design and naming of the suboptions (`enable`/`disable`, `no-colors`, `verbose-tree`)

quality concerns:

* [ ] Lukas: "it would be valuable to run `partest --neg --update-check` with the feature enabled and look at the diff"
  * this is something Dale and I could as part of our review process

followups, not blockers for this PR:

* [ ] integrate with or retire `-explaintypes`
* Lukas's comments about the wording of error messages:
  * [ ] "I find 'implicit error' very terse", can this be easily improved, is it a blocker?
  * [ ] ditto on the question at https://github.com/scala/scala/pull/7785#discussion_r539261029

dwijnand reviewed (commented):
> @dwijnand @SethTisue all upstream changes are now integrated!

That's excellent.  But I'm concerned.

> only one of those tests has different output with -Vimplicits enabled, I changed it accordingly.

This has my concern:

1. the other tests used to provide insight and now that's removed: why doesn't `-Vimplicits` provide the same or better diagnostics?
2. the one test that did produce output, why is it so less useful than the original?

I have a long, less important, list of questions/requests, around details of the code (formatting, using sys.env, using Option[String]/Option[Int], using try/catch NonFatal, creating pluginsNoImplicitFoundError, not hooking into pluginsNotifyImplicitSearch/Result) but I'm thinking it might be a quicker feedback if I try to apply those changes myself and find a way to make them into this PR (e.g. pushing to your branch or PRing your fork so we don't conflict).

SethTisue commented:
@tek Dale is off for the next two weeks, so progress on our end will slow down for a while (but we're now confident this can make 2.13.6)

tek commented:
damn, I did not see the notification for your comments, @dwijnand!

> This has my concern:
> 
>     1. the other tests used to provide insight and now that's removed: why doesn't `-Vimplicits` provide the same or better diagnostics?

`-Xlog-implicits` is a debugging-style option – it dumps a large amount of information that might not be useful. The original motivation for the plugin was to provide a more readable and minimized version of that.
The diagnostics are "better" only in the sense that they show what is most pertinent, it is not a replacement for `-Xlog-implicits` in that sense, which is why I asked if we really wanted to retire it.

>     2. the one test that did produce output, why is it so less useful than the original?

Not sure what you're referring to with "less useful"!

> using sys.env

oh right, I forgot about that. I originally added that to debug performance, but I think this can be removed.

> using try/catch NonFatal

yeah, I used that because I always saw a lot of potential for accessing some null attribute…since I'm only an amateur regarding the type encodings, maybe you can assess whether it's safe to remove it

> creating pluginsNoImplicitFoundError, not hooking into pluginsNotifyImplicitSearch/Result)

the motivation there is to allow a third-party compiler plugin to use the formatting engine to produce nice error messages for their libraries' specific errors. Are you suggesting that it should not exist or that it should replace the `Notify` hooks?

tek commented:
In response to @dwijnand and @lrytz, I see the following possible consequences for the issue with the old tracing messages:

1. Keep `-Xlog-implicits` for only those messages (or rename it)
2. Reuse another setting, like `-debug`
3. Output those messages when `-Vimplicits` is enabled (which should break formatting/UX of splain)
4. Remove those messages
5. Integrate messages into the splain UX

For 5., I would strongly suggest a followup, with intermittent 1.

dwijnand commented:
> [ ] Lukas: "it would be valuable to run `partest --neg --update-check` with the feature enabled and look at the diff"

I had a look at this and here's one simple example which looks reasonable to me:

```diff
-annotated-literal-annotation-arg.scala:15: error: bar
+annotated-literal-annotation-arg.scala:15: error: implicit error;
+!I e: Bar
+  bar
+
   implicitly[Bar]
             ^
```

Here's another case:

```diff
-sortedImplicitNotFound.scala:10: error: No implicit Ordering[Object] found to build a SortedSet[Object]. You may want to upcast to a Set[Int] first by calling `unsorted`.
+sortedImplicitNotFound.scala:10: error: implicit error;
+!I ev: Ordering[Object]
+  No implicit Ordering defined for Object.
+
+Ordering.ordered invalid because
+!I asComparable: Object => Comparable[_$2]
+  No implicit view available from Object => Comparable[_ >: Object].
+
+⋮
+Ordering.comparatorToOrdering invalid because
+!I cmp: Comparator[Object]
   ms.map(_ => o)
         ^
```

On the one hand this seems good to me: providing some useful details on discarded paths (hoping that it doesn't encourage more implicit view creation!!)

However it lost the `@implicitNotFound` message!  That seems to come from:

```scala
def map[B](f: A => B)(implicit @implicitNotFound(SortedSetOps.ordMsg) ev: Ordering[B]): CC[B]
```

(which I didn't even know worked.)

I feel convincible that that's not a blocker.  But then I found out that it looks like splain suppresses the messages on parents too:

```diff
-missing-implicit.scala:49: error: No F of Int
+missing-implicit.scala:49: error: implicit error;
+!I e: F[Int]
+  No F of Int
+
   implicitly[F[Int]]
             ^
-missing-implicit.scala:50: error: could not find implicit value for parameter e: M[Int] (No F of Int)
+missing-implicit.scala:50: error: implicit error;
+!I e: M[Int]
   implicitly[M[Int]]
             ^
```

`M` extends `F` but there's no "No F of Int" in the M case. 😞  Knowing how inheritance is used for typeclass hierarchies, I don't think we can afford this loss.

---

> > creating pluginsNoImplicitFoundError, not hooking into pluginsNotifyImplicitSearch/Result)
> 
> the motivation there is to allow a third-party compiler plugin to use the formatting engine to produce nice error messages for their libraries' specific errors. Are you suggesting that it should not exist or that it should replace the `Notify` hooks?

Sorry, I missed your question here, @tek!  No, what I meant was why it was creating its own hook rather than reusing the existing hooks and having splain be an AnalysisPlugin.  And, yeah, I don't think we should provide that plugin addition - at least not until it provides necessary.

> > using try/catch NonFatal
> 
> yeah, I used that because I always saw a lot of potential for accessing some null attribute…since I'm only an amateur regarding the type encodings, maybe you can assess whether it's safe to remove it

Sadly there is that potential, but I'd rather let it crash and fix the root causes that slowly sprinkle these all over.

> In response to @dwijnand and @lrytz, I see the following possible consequences for the issue with the old tracing messages:
> 
> 1. Keep `-Xlog-implicits` for only those messages (or rename it)
> 2. Reuse another setting, like `-debug`
> 3. Output those messages when `-Vimplicits` is enabled (which should break formatting/UX of splain)
> 4. Remove those messages
> 5. Integrate messages into the splain UX
> 
> For 5., I would strongly suggest a followup, with intermittent 1.

I'm not 100% sure, but I think unless we're convinced the messages are pointless or misleading, I dislike just removing them.  I'm also thinking we should keep `-Xlog-implicits` as an alias ("abbreviation") of `-Vimplicits`, so that means (1) and (3) clash.  I agree with you integrating it shouldn't block, so I'm thinking either just `-Vdebug` or `-Vimplicits -Vdebug`.

dwijnand committed:
Clean up splain

SethTisue commented:
I'm understanding for the first time, just now, that `-Vimplicits` doesn't just cause additional error information to be printed — it replaces the original error with something else. That's surprising to me, and Dale has identified some cases where not printing the original error means valuable information is lost.

Could it always print the original error, and then the splain stuff would come afterwards, as a supplement?

tek commented:
My suspicion is that both parameter messages and parent messages weren't around (or not at that location) when I initially wrote this five years ago.
The obvious solution is to copy the missing code over to splain, but I could also refactor it slightly so the annotation is shared with the default printer.

tek commented:
implemented the latter, please take a look. It's not complete though, this causes annotations to be ignored entirely for all but the first parameter in the chain.

tek commented:
@dwijnand I think that addresses all of the examples of information loss you quoted, do you have any others?

> Sorry, I missed your question here, @tek! No, what I meant was why it was creating its own hook rather than reusing the existing hooks and having splain be an AnalysisPlugin. And, yeah, I don't think we should provide that plugin addition - at least not until it provides necessary.

The existing hooks didn't include the necessary information, iirc. If they had, the plugin would have been a lot simpler :smile: 

Necessity wouldn't be an argument for this feature – its intended purpose is to allow people to provide a AnalyzerPlugin for e.g. `shapeless`, that includes additional formatting, like the `Record` rules that were part of the splain plugin and have now been removed. If you consider that to be not worth the addition of a new hook, I'll remove it.

> Sadly there is that potential, but I'd rather let it crash and fix the root causes that slowly sprinkle these all over.

Removed!

> I'm not 100% sure, but I think unless we're convinced the messages are pointless or misleading, I dislike just removing them. I'm also thinking we should keep `-Xlog-implicits` as an alias ("abbreviation") of `-Vimplicits`, so that means (1) and (3) clash. I agree with you integrating it shouldn't block, so I'm thinking either just `-Vdebug` or `-Vimplicits -Vdebug`.

ok!

dwijnand commented:
> @dwijnand I think that addresses all of the examples of information loss you quoted, do you have any others?

Thanks for the quick fix!  I'll re-run all the tests with the option enabled again (and actually we should just have that commit run in the PR).

> Necessity wouldn't be an argument for this feature – its intended purpose is to allow people to provide a AnalyzerPlugin for e.g. `shapeless`, that includes additional formatting, like the `Record` rules that were part of the splain plugin and have now been removed. If you consider that to be not worth the addition of a new hook, I'll remove it.

As it's simple (and already implemented) I'm fine with keeping it, you've convinced me 😄

I have a patch on top of yours which applies a whole bunch of changes, which allowed me to understand the code, and try different things.  I'll look to rebasing it onto your branch here and PRing your fork next week so you have a look.

tek commented:
> we should just have that commit run in the PR

sorry, don't understand what that means. do you need me to do something there?

> As it's simple (and already implemented) I'm fine with keeping it, you've convinced me smile

:slightly_smiling_face: 

> I have a patch on top of yours which applies a whole bunch of changes, which allowed me to understand the code, and try different things. I'll look to rebasing it onto your branch here and PRing your fork next week so you have a look.

ok cool. do you want me to squash again now?

dwijnand commented:
> > we should just have that commit run in the PR
> 
> sorry, don't understand what that means. do you need me to do something there?

My bad, that was a quick self/team-reminder.  See https://github.com/scala/scala/pull/9557 for an example: make a change with the flag default on, to get a CI-verified test pass commit, and then default it off in the next commit for the merge.

> ok cool. do you want me to squash again now?

No, hopefully we can get my changes in and then we can squash it all up (I've been debating whether it's best to have my changes after or squashed with yours - I don't care about the commit credit, I'm just thinking if any of the two options are preferable.)

tek commented:
> My bad, that was a quick self/team-reminder. See #9557 for an example: make a change with the flag default on, to get a CI-verified test pass commit, and then default it off in the next commit for the merge.

ahhh clever!

> No, hopefully we can get my changes in and then we can squash it all up (I've been debating whether it's best to have my changes after or squashed with yours - I don't care about the commit credit, I'm just thinking if any of the two options are preferable.)

:+1:

tek commented:
so, we decided to remove the link to the docs.scala-lang.org link in favor of a better description for the option. what are the requirements for this?

dwijnand commented:
https://github.com/scala/docs.scala-lang/pull/1840 is the docs page, which needs an update with how I've reworked the options.  Previously everything was under a single "MultiChoiceSetting", while now it's broken up, which gives more lines to give some basic description, and have that docs page nicely provide more context and nice examples.

Personally I welcome suggestions on improved description, but I'm overly concerned if we were to ship what we have currently in the PR.

tek commented:
updated the doc page, please let me know what's bad!

dwijnand committed:
Default -Vimplicits to true to see test impact

dwijnand committed:
Revert -Vimplicits back to default false

SethTisue reviewed (approved):
(final review/merge by @lrytz?)

lrytz reviewed (approved):
LGTM too!

The message formatting / wording can still be improved, that would be helpful. Basically what's discussed in these comments: https://github.com/scala/scala/pull/7785/files#diff-0d00a887e46a40735249285fb63ea04a8aa3de82943535fdfdbf86106cdcb88d

tek commented:
wooooo :partying_face: :tada:

kubukoz commented:
WOOOOOOOOOOO!

bl-ue commented:
So does this close https://github.com/tek/splain/issues/4?

tek commented:
fuck yeah!

tribbloid commented:
At this moment (Sept 2021), who is the maintainer/reviewer of this feature?

There are a few bugfixes in splain 0.5.9 that wasn't included.

Unfortunately the first patch to scala compiler takes a lot of copy and paste. IMHO, we can't afford to do this repeatedly for hotfixes.

After consulting with @tek. I may have to rebase them on this repo

SethTisue commented:
> who is the maintainer/reviewer of this feature?

any of us, all of us. PRs welcome

tribbloid commented:
@SethTisue thanks a lot. I've talked to @tek and discovered that many config options are also omitted:

https://github.com/tek/splain/pull/58

So I'll have to introduce them first. The following 2 PR should be ideally submitted in succession:

1. add the following configurations that are ingested though `-P:splain:<param>[:<value>]` options:

```scala
  val keyAll = "all"
  val keyImplicits = "implicits"
  val keyFoundReq = "foundreq"
  val keyInfix = "infix"
  val keyBounds = "bounds"
  val keyColor = "color"
  val keyBreakInfix = "breakinfix"
  val keyCompact = "compact"
  val keyTree = "tree"
  val keyBoundsImplicits = "boundsimplicits"
  val keyTruncRefined = "truncrefined"
  val keyRewrite = "rewrite"
  val keyKeepModules = "keepmodules"

  val analyzer =
    new { val global = SplainPluginCompat.this.global } with Analyzer {
      def featureImplicits = boolean(keyImplicits)
      def featureFoundReq = boolean(keyFoundReq)
      def featureInfix = boolean(keyInfix)
      def featureBounds = boolean(keyBounds)
      def featureColor = boolean(keyColor)
      def featureBreakInfix = int(keyBreakInfix).filterNot(_ == 0)
      def featureCompact = boolean(keyCompact)
      def featureTree = boolean(keyTree)
      def featureBoundsImplicits = boolean(keyBoundsImplicits)
      def featureTruncRefined = int(keyTruncRefined).filterNot(_ == 0)
      def featureRewrite = opt(keyRewrite, "")
      def featureKeepModules = int(keyKeepModules)
    }
```
(feel free to suggest alternative names)

2. backport the patch

Wonder if I can catch the 2.13.7 release deadline?
